### PR TITLE
[Feat] Added Mixup data augmentation

### PR DESCRIPTION
**Pull request by shijianjian**
### Description
The implementation for the paper https://arxiv.org/pdf/1710.09412.pdf

It should be a useful feature while it includes a label altering as well. May discuss further on the API signatures, etc.

Similar logic also includes:
CutMix: https://arxiv.org/abs/1905.04899

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Comment by shijianjian** [comment]
I do not think this part is ready to merge since the API usage is a bit way off what we already have.

Currently, I named it as ```label_soften_augmentation``` since it augmented not only images but the label as well. I am not sure what is the most desired way to use this. Still need some discussion first.

File: kornia/geometry/transform/crop.py [hunk 1]
```diff
@@ -201,29 +208,67 @@ def _infer_bounding_box(boxes: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor
...
-    assert len((boxes[:, 1, 0] - boxes[:, 0, 0] + 1).unique()) == 1, \
-        "Boxes can only have one widths, got %s" % str((boxes[:, 1, 0] - boxes[:, 0, 0] + 1).unique())
-    assert len((boxes[:, 2, 1] - boxes[:, 0, 1] + 1).unique()) == 1, \
-        "Boxes can only have one heights, got %s" % str((boxes[:, 2, 1] - boxes[:, 0, 1] + 1).unique())
 
-    width: torch.Tensor = (boxes[:, 1, 0] - boxes[:, 0, 0] + 1)[0]
-    height: torch.Tensor = (boxes[:, 2, 1] - boxes[:, 0, 1] + 1)[0]
-    return (height, width)
+
+def bbox_to_mask(boxes: torch.Tensor, width: int, height: int) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/geometry/transform/crop.py:245]: this needs proper documentation. Not included to sphinx docs.

Comment (edgarriba) [reply kornia/geometry/transform/crop.py:245]: Besides, I'm thinking since we start to have few functionalities around bounding boxes manipulation, we could have a small sub-module dedicated to that: `kornia.geometry.transform.boxes` ? /cc @shijianjian @ducha-aiki @lferraz 
I think @lferraz had some suggestions about this

File: test/augmentation/test_augmentation_mix.py [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+
+class TestRandomMixUp:
+
+    def smoke_test(self, device):
+        f = RandomMixUp()
+        repr = "RandomMixUp(p=1.0, max_lambda=tensor(1.0), same_on_batch=False)"
+        assert str(f) == repr
+
+    def test_random_mixup_p1(self, device):
```
Comment (edgarriba) [inline test/augmentation/test_augmentation_mix.py:22]: you should inclide `dtype` too

File: test/augmentation/test_augmentation_mix.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        torch.manual_seed(0)
+        f = RandomMixUp()
+
+        input = torch.stack([torch.ones(1, 3, 4), torch.zeros(1, 3, 4)]).to(device)
+        label = torch.tensor([1, 0]).to(device)
```
Comment (edgarriba) [inline test/augmentation/test_augmentation_mix.py:27]: recommended way to initialize tensor is `torch.tensor([1, 0], device=device, dtype=dtype)` since it directly does initialization on the specified device and type.

**Review by edgarriba — CHANGES_REQUESTED** [review]
Few more comments.

File: kornia/augmentation/functional.py [hunk 4]
```diff
@@ -908,3 +910,122 @@ def apply_equalize(input: torch.Tensor, params: Dict[str, torch.Tensor]) -> torc
...
+    lam = params['mixup_lambdas'].view(-1, 1, 1, 1).expand_as(input).to(input.device)
+    inputs = input * (1 - lam) + input_permute * lam
+    labels = torch.stack([labels.float(), labels_permute.float(), params['mixup_lambdas'].to(labels.device)], dim=-1)
+    return inputs, labels
+
+
+def apply_cutmix(input: torch.Tensor, labels: torch.Tensor,
+                 params: Dict[str, torch.Tensor]) -> Tuple[torch.Tensor, torch.Tensor]:
+    r"""Apply cutmix to images in a batch. CutMix augmentation strategy: patches are cut and pasted
+    among training images where the ground truth labels are also mixed proportionally to the area of the patches
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/augmentation/functional.py:969]: What do you mean to "link here to the module version"? You mean kornia module version?

File: kornia/filters/kernels.py [hunk 5]
```diff
@@ -478,14 +478,20 @@ def get_motion_kernel2d(kernel_size: int, angle: Union[torch.Tensor, float],
         raise TypeError("ksize must be an odd integer >= than 3")
 
     if not isinstance(angle, torch.Tensor):
-        angle = torch.tensor(angle)
+        angle = torch.tensor([angle])
 
-    assert angle.dim() == 0, f"angle must be a 0-dim tensor. Got {angle}."
+    angle = cast(torch.Tensor, angle)
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/filters/kernels.py:483]: Yes, but the auto-detected typing was still "float/tensor" if exclude this line. Just to secure the types here.

File: kornia/geometry/transform/crop.py [hunk 6]
```diff
@@ -178,9 +183,11 @@ def crop_by_boxes(tensor, src_box, dst_box,
     # simulate broadcasting
     dst_trans_src = dst_trans_src.expand(tensor.shape[0], -1, -1).type_as(tensor)
 
-    bbox = _infer_bounding_box(dst_box)
+    bbox = infer_box_shape(dst_box)
+    assert len(bbox[0].unique()) == 1 and len(bbox[1].unique()) == 1, "cropping height and width" \
```
[the remark this answers is not in the document]
Comment (shijianjian) [reply kornia/geometry/transform/crop.py:187]: I will refine it in this PR.

File: kornia/augmentation/mix_augmentation.py [hunk 7]
```diff
@@ -71,7 +71,7 @@ def acc_mixup(y, logits):
 
     Args:
         p (float): probability for performing mixup. Default is 0.5.
-        max_lambda (float or torch.Tensor, optional): max value of mixup strength. Default is 1.
+        lam (float or torch.Tensor, optional): min-max value of mixup strength. Default is 0-1.
```
Comment (edgarriba) [inline kornia/augmentation/mix_augmentation.py:74]: why not `lambda_val` ?
